Add prompt implementation for service methods lock & unlock

warusadura opened the pull request:
Please see the commits.

bilelmoussaoui commented on client/src/crypto/native.rs:
```diff
@@ -81,6 +81,17 @@ pub(crate) fn generate_public_key(private_key: impl AsRef<[u8]>) -> Vec<u8> {
 
+pub(crate) fn generate_public_key_for_secret_exchange(private_key: impl AsRef<[u8]>) -> Vec<u8> {
+    let private_key_uint = BigUint::from_bytes_be(private_key.as_ref());
+    let dh_generator = {
```
Should use LazyLock nowadays.

warusadura replied:
Thanks, done.

bilelmoussaoui commented on client/src/crypto/native.rs:
```diff
@@ -110,6 +121,35 @@ pub(crate) fn generate_aes_key(
+
+    let mut common_secret_bytes = common_secret.to_bytes_be();
+    let mut common_secret_padded = vec![0; 192 - common_secret_bytes.len()];
+    // inefficient, but ok for now
```
what is inefficient about it

warusadura replied:
I was just following the existing implementations.

bilelmoussaoui commented on client/src/crypto/native.rs:
```diff
@@ -233,3 +273,42 @@ fn powm(base: &BigUint, mut exp: BigUint) -> BigUint {
+fn powm_for_secret_exchange(base: &BigUint, mut exp: BigUint) -> BigUint {
+    let modulus = {
+        // for key exchange
+        static DH_PRIME: OnceLock<BigUint> = OnceLock::new();
```
LazyLock

bilelmoussaoui commented on client/src/crypto/openssl.rs:
```diff
@@ -94,6 +94,19 @@ pub(crate) fn generate_public_key(private_key: impl AsRef<[u8]>) -> Vec<u8> {
 
+pub(crate) fn generate_public_key_for_secret_exchange(private_key: impl AsRef<[u8]>) -> Vec<u8> {
+    let private_key_bn = BigNum::from_slice(private_key.as_ref()).unwrap();
+    Dh::from_pqg(
+        BigNum::get_rfc3526_prime_1536().unwrap(),
+        None,
+        BigNum::from_u32(2).unwrap(),
+    )
+    .and_then(|key| key.set_private_key(private_key_bn))
+    .unwrap()
```
See my latest PR, don't unwrap here

warusadura replied:
Implemented, thanks!

bilelmoussaoui commented on client/src/crypto/openssl.rs:
```diff
@@ -127,6 +140,39 @@ pub(crate) fn generate_aes_key(
+        None,
+        BigNum::from_u32(2).unwrap(),
+    )
+    .and_then(|key| key.set_private_key(private_key_bn))
```
Same here

bilelmoussaoui commented on client/src/crypto/openssl.rs:
```diff
@@ -127,6 +140,39 @@ pub(crate) fn generate_aes_key(
+    .unwrap();
+
+    let mut common_secret_padded = vec![0; 192 - common_secret_bytes.len()];
+    // inefficient, but ok for now
```
Same here

bilelmoussaoui commented on server/Cargo.toml:
```diff
@@ -9,6 +9,7 @@ rust-version.workspace = true
 version.workspace = true
 
 [dependencies]
+base64 = "0.22.1"
```
```suggestion
base64 = "0.22"
```

warusadura replied:
Done.

bilelmoussaoui commented on server/src/main.rs:
```diff
@@ -2,6 +2,9 @@ mod collection;
 mod item;
 mod prompt;
+mod prompter;
+mod prompter_callback;
+mod secret_exchange;
```
Those two seems to be specific to the gnome prompting mechanism, so probably should go under a gnome sub-module.

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/prompt.rs:
```diff
@@ -1,19 +1,40 @@
-        todo!()
+    pub async fn prompt(
+        &self,
+        _window_id: &str,
```
This should use a ashpd::WindowIdentifierType here

bilelmoussaoui replied:
i know it is not used by the gnome implementation but it will eventually used by something better.

bilelmoussaoui commented on server/src/prompter.rs:
```diff
@@ -0,0 +1,29 @@
+    fn perform_prompt(
+        &self,
+        callback: OwnedObjectPath,
+        type_: &str, // 'password' or 'confirm', put this in Enum?
```
Yes, this has to be turned into an enum to avoid random strings being passed.

warusadura replied:
Done.

bilelmoussaoui commented on server/src/prompter.rs:
```diff
@@ -0,0 +1,29 @@
+        &self,
+        callback: OwnedObjectPath,
+        type_: &str, // 'password' or 'confirm', put this in Enum?
+        properties: HashMap<&str, OwnedValue>,
```
I assume the properties are fixed/known dictionnaries, if so, they must be defined as structs.

bilelmoussaoui commented on server/src/prompter.rs:
```diff
@@ -0,0 +1,29 @@
+        callback: OwnedObjectPath,
+        type_: &str, // 'password' or 'confirm', put this in Enum?
+        properties: HashMap<&str, OwnedValue>,
+        exchange: &str,
```
What is the content of this? is it really a string?

warusadura replied:
Yes. The content:
```
'[sx-aes-1]\npublic=GfG2V4/Q/IgZpA9mV3QLnmFiUBdHSwIaNkZJpn//vuQLn1P2ujRT03ZBOJpKZsq/IYnWbTDfuRDue+kq7cWlp6KEdsv+6ZM57kWE7V4UnBcEtiuZB/DwuGmVL1JaKdqWbs9AVN3D+l6ZFrYek0WZM2YcOiGWUUD7TzhUai6RFVkgKC+KC3BQpldDefZCb5jxiYldC6gRy3+AQBnkpRqlT0eeh/9ecL4FieiRhfYJpUmooMUQY9Y7fFycVbl5aGw5\n'
```

bilelmoussaoui commented on server/src/prompter.rs:
```diff
@@ -0,0 +1,29 @@
+};
+
+#[proxy(
+    default_service = "org.gnome.keyring.SystemPrompter",
```
gnome specific -> gnome module

warusadura replied:
Done.

bilelmoussaoui commented on server/src/prompter_callback.rs:
```diff
@@ -0,0 +1,47 @@
+
+impl PrompterCallback {
+    pub async fn new(service: Service) -> Self {
+        let index = service.session_index().await;
```
session or prompt_index?

warusadura replied:
Fixed this, thanks!

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+
+// Initial C implementation: https://gitlab.gnome.org/GNOME/gcr/-/blob/master/gcr/gcr-secret-exchange.c
+
+// The initial implementation of SecretExchange/GCRSecretExchange uses a KeyFile
+// to encode/parse the payload. And in this implementation the payload is based
+// on a HashMap.
```
```suggestion
// The initial implementation of SecretExchange/GCRSecretExchange uses a KeyFile
// to encode/parse the payload. In this implementation the payload is based
// on a HashMap.
```

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+const CIPHER_TEXT_LEN: usize = 16;
+
+#[derive(Debug)]
+pub struct SecretExchange {
```
Zeroize the type.

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+
+    // Creates the shared secret: an AES key
+    pub fn create_shared_secret(&self, exchange: &str) -> String {
+        let decoded = decode(exchange).unwrap();
```
Avoid unwrap, use expect if the error is never supposed to happen with a comment why it is not

warusadura replied:
Yeah, thanks!

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+// Retrieves the secret from final secret exchange string
+pub(crate) fn retrieve_secret(exchange: &str, aes_key: &str) -> Option<Zeroizing<Vec<u8>>> {
+    let decoded = decode(exchange).unwrap();
+    let secret = decoded.get(SECRET);
+
+    // If we cancel an ongoing prompt call, the final exchange won't have the secret
+    // or IV. The following is to avoid `Option::unwrap()` on a `None` value
+    secret?;
+
+    let secret = secret.unwrap();
```
```suggestion
    // If we cancel an ongoing prompt call, the final exchange won't have the secret
    // or IV. The following is to avoid `Option::unwrap()` on a `None` value
    let secret = decoded.get(SECRET)?;

```

warusadura replied:
Done.

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+    let secret = secret.unwrap();
+    if secret.len() != CIPHER_TEXT_LEN {
+        // to avoid a short secret/cipher-text causing an UnpadError during decryption
+        let false_secret: Vec<u8> = vec![0, 1];
+        return Some(Zeroizing::new(false_secret));
```
```suggestion
        let false_secret = vec![0, 1];
        return Some(Zeroizing::new(false_secret));
```

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+        return Some(Zeroizing::new(false_secret));
+    }
+
+    let iv = decoded.get(IV).unwrap();
+    let decoded = decode(aes_key).unwrap();
+    let aes_key = Key::new(decoded.get(PRIVATE).unwrap().to_vec());
```
```suggestion
    let iv = decoded.get(IV)?;
    let decoded = decode(aes_key)?;
    let aes_key = Key::new(decoded.get(PRIVATE)?.to_vec());
```

bilelmoussaoui commented on server/src/secret_exchange.rs:
```diff
@@ -0,0 +1,131 @@
+    use super::*;
+
+    #[test]
+    fn test_secret_exchange() {
```
should the opposite also be tested?

bilelmoussaoui commented on server/src/service.rs:
```diff
@@ -114,8 +115,16 @@ impl Service {
-        let (unlocked, _not_unlocked) = self.set_locked(false, &objects).await?;
+        let (unlocked, not_unlocked) = self.set_locked(false, &objects).await?;
+        if !not_unlocked.is_empty() {
+            let prompt = Prompt::new(self.clone()).await;
```
The prompt also needs a way to pass it a future that would be executed when the prompt is done

warusadura replied:
Addressed [here](https://github.com/bilelmoussaoui/oo7/pull/173#discussion_r1966813203)

bilelmoussaoui reviewed (commented):
also needs a rebase

warusadura commented:
Thanks for the prompt review @bilelmoussaoui

bilelmoussaoui commented on client/src/crypto/native.rs:
```diff
@@ -1,6 +1,6 @@
 use std::{
     ops::{Mul, Rem, Shr},
-    sync::LazyLock,
+    sync::{LazyLock, OnceLock},
```
This lacks an openssl implementation as well as that is what most distros will likely use the daemon with

warusadura replied:
Will do, thanks!

warusadura commented:
Sorry for the delay @bilelmoussaoui 

I have removed the prompt implementation for now. Just wanted get the crypto part and secret_exchange impl cleared before moving into the Prompt impl. I have removed a lot of unwraps, but I'm still not sure this is any good. Very much like to hear what you think of the current impl :)

bilelmoussaoui commented on server/src/gnome/secret_exchange.rs:
```diff
@@ -0,0 +1,174 @@
+}
+
+// Retrieves the secret from final secret exchange string
+pub(crate) fn retrieve_secret(exchange: &str, aes_key: &str) -> Option<Zeroizing<Vec<u8>>> {
```
This should probably return a oo7::Secret.

warusadura replied:
Yeah, it should. Done.

bilelmoussaoui commented on server/src/gnome/secret_exchange.rs:
```diff
@@ -0,0 +1,174 @@
+const CIPHER_TEXT_LEN: usize = 16;
+
+#[derive(Debug, Zeroize)]
+pub struct SecretExchange {
```
I am still not sure how the SecretExchange is supposed to be used. The current API, even if used internally only, is not really nice. why there is a need for a SecretExchange struct at all? who would call begin/create_shared_secret? and what about retrieve_secret? I would need to see how it is used in a real-case scenario to provide any meaningful feedback

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+    fn perform_prompt(
+        &self,
+        callback: OwnedObjectPath,
+        type_: &str,
+        properties: HashMap<&str, OwnedValue>,
```
Type safety, like mentioned before.

warusadura replied:
Done, thanks!
Quick question: why using a HashMaps isn't type safe? I couldn't find an answer for this from the zbus book.

bilelmoussaoui replied:
What i meant here is that you can pass an arbitrary dict when in reality it is a struct but as the only way to pass that over dbus is to use a dict, we still shouldn't use a dict from Rust.

warusadura replied:
Understood, thanks and sorry for the delay.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+impl PrompterCallback {
+    pub async fn prompt_ready(
+        &self,
+        reply: &str,
```
What the reply contains?

warusadura replied:
'yes' or 'no'.

bilelmoussaoui replied:
Make it an enum then.

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+    pub async fn prompt_ready(
+        &self,
+        reply: &str,
+        properties: HashMap<&str, OwnedValue>,
```
Type safety, like above

bilelmoussaoui replied:
This was marked as resolved yet I still see properties being `HashMap<&str, Value>` instead of the newly added struct

bilelmoussaoui replied:
also the properties passed are never used, how come?

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+            PromptRole::Lock => {
+                if properties.is_empty() {
+                    // First PromptReady call
+                    let mut properties: HashMap<&str, zvariant::OwnedValue> = HashMap::new();
```
Same as above

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+
+impl PrompterCallback {
+    pub async fn new(service: Service) -> Self {
+        let index = service.prompt_index().await;
```
You should create a PrompterProxy here and keepe it around for the functions below.

warusadura replied:
Good idea. But I can't access `self` inside `tokio::spawn`. All these following functions are associated functions.

bilelmoussaoui commented on server/src/prompt.rs:
```diff
@@ -1,25 +1,92 @@
+};
+
+#[derive(Debug, Clone)]
+pub enum PromptRole {
+    Lock,
+}
```
What else this enum might contain? fill it with other types and add a todo! for the match rule

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+                } else {
+                    // Second PromptReady call with the final exchange
+                    let service = self.service.clone();
+                    let objects = prompt.objects().clone();
+                    let result = Value::new(&objects).try_to_owned().unwrap();
+
+                    tokio::spawn(async move {
+                        let _ = service.set_locked(true, &objects, true).await;
+                    });
```
Instead of passing the objects, this should pass a future directly that you can spawn here. Keeping the logic that is supposed to be in Service::set_locked, stay there.

warusadura replied:
https://github.com/bilelmoussaoui/oo7/pull/173#discussion_r1966813203

bilelmoussaoui commented on server/src/service.rs:
```diff
@@ -410,4 +433,23 @@ impl Service {
+        n_prompts
+    }
+
+    pub async fn prompt(&self) -> Option<Prompt> {
+        let prompts = self.prompts.lock().await;
+
+        prompts.last().cloned()
+    }
+
+    pub async fn remove_prompt(&self) {
+        // prompts should always contain one item during a prompt related operation and
+        // it should be cleaned up afterwards.
+        self.prompts.lock().await.pop();
+    }
```
If there would always be one single prompt (which i don't believe is true, except for maybe GNOME specifically) as the API clearly allows modal keyring dialogs so every app could have an ongoing prompt operation.

warusadura replied:
Yeah, this is correct and I have removed this restriction.

bilelmoussaoui commented on server/src/service.rs:
```diff
@@ -318,6 +336,7 @@ impl Service {
         &self,
         locked: bool,
         objects: &[OwnedObjectPath],
+        prompt: bool,
```
why is this needed?

warusadura replied:
https://github.com/bilelmoussaoui/oo7/pull/173#discussion_r1966813203

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,205 @@
+                    let connection = connection.clone();
+                    let path = self.path.clone();
+
+                    tokio::spawn(PrompterCallback::perform_prompt(
+                        connection, path, "confirm", properties, exchange,
+                    ));
```
Not sure why this has to be in PrompterCallback,

warusadura replied:
What do you mean? Maybe the confusion is caused by the fact that I'm using the same name  as `Prompter.perform_prompt`. `PrompterCallback::perform_prompt` is just a wrapper calling `Prompter.perform_prompt` with other code to avoid code duplication.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,226 @@
+                    })?;
+                    let exchange = secret_exchange.begin();
+
+                    let properties = Properties {
```
As those properties are static, move them to new constructors in Properties impl. Like, for_lock, for_unlock and so on.

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,226 @@
+                        description: "Confirm locking \"login\", Keyring.".to_string(),
+                        message: "Lock Keyring".to_string(),
+                        caller_window: "".to_string(),
+                        warning: "".to_string(),
```
Optional ones should be `Option<String>`.

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,226 @@
+        match prompt.role() {
+            PromptRole::Lock => {
+                if properties.is_empty() {
+                    // First PromptReady call
```
Shouldn't most of the prompting code be exactly the same no matter the PromptRole? the only thing that i see being different is the Properties and the future to execute once the prompt is done

warusadura replied:
I think this is correct. I have just pushed the unlock commit (wip) and the code looks mostly the same with a couple of changes.

warusadura commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,287 @@
+#[derive(Debug, DeserializeDict, SerializeDict, Type)]
+#[zvariant(signature = "a{sv}")]
+// System prompt properties
+pub struct Properties {
```
With this approach, during the second PromptReady call return I'm getting the following error, `Unexpected non-0 padding byte 1`. I can't figure out why. I think this is related to [this](https://gitlab.freedesktop.org/dbus/zbus/-/issues/82). Then again all the fields here are properly renamed using `#[zvariant(rename = "")`

bilelmoussaoui replied:
Check what gets send over the bus with busctl and see what is going on? Probably a mismatched type

warusadura replied:
Will do, thanks.

warusadura replied:
I tried everything I can think of. I think this is happening due to a bug in zbus. Probably related to zbus issues #1193 and #320.

bilelmoussaoui replied:
Can you write a minimal reproducer of the issue? I am using the exact same thing in ashpd and it works fine.

warusadura replied:
Will do.

> I am using the exact same thing in ashpd and it works fine.

Yeah, I did see that.

bilelmoussaoui replied:
Actually the issues you mentioned are about the OwnedValue/Value derive macros which you don't use here. So the issue is very likely in the types you are using inside Properties struct. Maybe they are all passed as strings?

warusadura replied:
So [these](https://gitlab.gnome.org/GNOME/gcr/-/blob/main/gcr/gcr-prompt.c#L95) are the GcrPrompt properties.

bilelmoussaoui replied:
Show me the dbus traffic instead.

warusadura replied:
https://upload.disroot.org/r/uCEPb8Pu#Z6OOnI1W4Fx5uALHyb8s5IIVaaDAngKl5tPm5m0tk04=

bilelmoussaoui replied:
Can you confirm this is caused by the properties and not the the reply param of PromptReady?

bilelmoussaoui replied:
Then keep the Properties struct there and add a todo comment about it so we can investigate it further.

warusadura replied:
got it, thanks! I think I'll have to keep this diff as well. I can't test anything without it :)

warusadura replied:
Just locally.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,287 @@
+    description: Option<String>,
+    message: Option<String>,
+    #[zvariant(rename = "caller-window")]
+    caller_window: Option<String>,
```
This should be a WindowIdentifierType from ashpd

warusadura replied:
Don't know anything about ashpd. I'll look into it :)

bilelmoussaoui replied:
You don't have to do anything special here, just use that type.

warusadura replied:
hey. I tried this and I get the following error: `the trait bound `WindowIdentifierType: Serialize` is not satisfied`

bilelmoussaoui replied:
easy to add upstream, will take care of that :-)

warusadura replied:
Thanks :)

warusadura replied:
Thanks!

warusadura replied:
https://github.com/bilelmoussaoui/oo7/pull/173#discussion_r1876066717

I tried this (also the above) and now I'm getting the following error,
![Screenshot From 2025-02-18 15-19-36](https://github.com/user-attachments/assets/120915ed-f3e4-4107-9709-39ab08279ab1)

warusadura replied:
![image](https://github.com/user-attachments/assets/50bf5b9e-9e5d-4c67-b407-8c9f77dacf8f)

A6GibKm replied:
Please paste here the windows id that was used.

warusadura replied:
```
diff --git a/Cargo.lock b/Cargo.lock
index 9969255..e35cb09 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1224,6 +1224,7 @@ dependencies = [
 name = "oo7-daemon"
 version = "0.4.0"
 dependencies = [
+ "ashpd",
  "base64",
  "clap",
  "enumflags2",
diff --git a/server/Cargo.toml b/server/Cargo.toml
index dd20f6b..d677d4f 100644
--- a/server/Cargo.toml
+++ b/server/Cargo.toml
@@ -9,6 +9,7 @@ rust-version.workspace = true
 version.workspace = true
 
 [dependencies]
+ashpd = { workspace = true, features = ["backend"] }
 base64 = "0.22"
 clap.workspace = true
 enumflags2 = "0.7"
diff --git a/server/src/gnome/prompter.rs b/server/src/gnome/prompter.rs
index da3fda3..4f66e55 100644
--- a/server/src/gnome/prompter.rs
+++ b/server/src/gnome/prompter.rs
@@ -3,6 +3,7 @@
 
 use std::sync::Arc;
 
+use ashpd::WindowIdentifierType;
 use clap::error::Result;
 use oo7::dbus::ServiceError;
 use serde::{Deserialize, Serialize};
@@ -40,7 +41,7 @@ pub struct Properties {
     #[zvariant(rename = "choice-chosen")]
     choice_chosen: Option<bool>,
     #[zvariant(rename = "caller-window")]
-    caller_window: Option<String>,
+    caller_window: Option<WindowIdentifierType>,
     #[zvariant(rename = "continue-label")]
     continue_label: Option<String>,
     #[zvariant(rename = "cancel-label")]
@@ -48,7 +49,7 @@ pub struct Properties {
 }
 
 impl Properties {
-    fn for_lock(object: &str, window_id: &str) -> Self {
+    fn for_lock(object: &str, window_id: &WindowIdentifierType) -> Self {
         Self {
             title: None,
             message: Some("Lock Keyring".to_owned()),
@@ -58,13 +59,13 @@ impl Properties {
             password_strength: None,
             choice_label: None,
             choice_chosen: None,
-            caller_window: Some(window_id.to_owned()),
+            caller_window: Some(window_id.clone()),
             continue_label: Some("Lock".to_owned()),
             cancel_label: Some("Cancel".to_owned()),
         }
     }
 
-    fn for_unlock(object: &str, warning: &str, window_id: &str) -> Self {
+    fn for_unlock(object: &str, warning: &str, window_id: &WindowIdentifierType) -> Self {
         Self {
             title: Some("Unlock Keyring".to_owned()),
             message: Some("Authentication required".to_owned()),
@@ -77,7 +78,7 @@ impl Properties {
             password_strength: None,
             choice_label: None,
             choice_chosen: None,
-            caller_window: Some(window_id.to_owned()),
+            caller_window: Some(window_id.clone()),
             continue_label: Some("Unlock".to_owned()),
             cancel_label: Some("Cancel".to_owned()),
         }
@@ -146,7 +147,7 @@ pub trait Prompter {
 // org.gnome.keyring.internal.Prompter.Callback
 
 pub struct PrompterCallback {
-    window_id: String,
+    window_id: WindowIdentifierType,
     // SecretExchange public_key
     secret_exchange_pub_key: Arc<RwLock<String>>,
     // SecretExchange aes_key
@@ -215,13 +216,13 @@ impl PrompterCallback {
 }
 
 impl PrompterCallback {
-    pub async fn new(window_id: &str, service: Service) -> Self {
+    pub async fn new(window_id: WindowIdentifierType, service: Service) -> Self {
         let index = service.prompt_index().await;
         Self {
-            window_id: window_id.to_owned(),
             secret_exchange_pub_key: Default::default(),
             secret_exchange_aes_key: Default::default(),
             path: OwnedObjectPath::try_from(format!("/org/gnome/keyring/Prompt/p{index}")).unwrap(),
+            window_id,
             service,
         }
     }
diff --git a/server/src/prompt.rs b/server/src/prompt.rs
index ca48004..e8cdd57 100644
--- a/server/src/prompt.rs
+++ b/server/src/prompt.rs
@@ -32,7 +32,7 @@ pub struct Prompt {
 impl Prompt {
     pub async fn prompt(
         &self,
-        window_id: &str,
+        window_id: ashpd::WindowIdentifierType,
         #[zbus(connection)] connection: &zbus::Connection,
         #[zbus(object_server)] object_server: &zbus::ObjectServer,
     ) -> Result<(), ServiceError> {

```

bilelmoussaoui replied:
the diff you pasted doesn't include the call to PromptCallback::new which probably is passing the wrong WindowIdentifierType. Instead, make those calls pass Option<T> and pass None when you don't have a window identifier like in GNOME instead of passing Some("") which is wrong.

warusadura replied:
Thanks for the reply.
No, the diff there. And there is no `Some("")` in this pr anymore.
So, the `window_id` parameter is coming from the Prompt method `prompt`. We don't pass this. This method get called when we create and serve a Prompt on the object tree. All the time the value of this parameter is `''`. And we can't use `window_id: Option<ashpd::WindowIdentifierType>`. This causes "Signature mismatch: got `s`, expected `ms`" error.

bilelmoussaoui replied:
You can't use Option<T> obviously, like I mentioned previously for the prompt reply, it must be Optional<T>

warusadura replied:
Tried `Optional` too :) "the trait bound `WindowIdentifierType: Default` is not satisfied"

bilelmoussaoui replied:
anyways at this point focus at finishing the basics of the implementation, i will take care of cleaning things afterwards.

warusadura replied:
Understood, thanks! Will keep the `window_id: &str` for now (easy for testing and compiling). I'll add a comment regarding WindowIdentifierType.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,287 @@
+pub enum Reply {
+    No,
+    Yes,
+    Empty,
```
What does empty corresponds to? Please link in the code to where those answers are defined

warusadura replied:
Empty corresponds to: `''`. The first PromptReady call's `reply` argument is always `''` and also the `properties` dict is empty.

![image](https://github.com/user-attachments/assets/e0c12bcd-c794-43b2-bc4e-d643d518ac68)

bilelmoussaoui replied:
Check if you can use the zvariant::Optional type here and drop the Empty variant.

warusadura replied:
Will do, thanks!

warusadura replied:
hey. I tried `reply: Optional<Reply>` approach and I think the current approach with Empty variant is bit better or shorter. Because all we have to do is impl Serialize and Deserialize. When using Optional, I had to impl `NoneValue` and bunch of String, &str traits. I just stopped at: trait `FromStr` is not implemented for `Reply`. Should I keep `Empty` variant or continue with `Optional`?

bilelmoussaoui replied:
Use zvariant::Optional please. and you shouldn't have to implement that many traits but if you are struggling with it, you can leave it as is and i will take care of it before merging.

warusadura replied:
Ok, thanks :)

warusadura replied:
Done.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,287 @@
+    Empty,
+}
+
+const NO: &str = "no";
```
No need for those consts

warusadura replied:
Removed, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,287 @@
+#[derive(Deserialize, Serialize, Debug, Type)]
+#[serde(rename_all = "lowercase")]
+#[zvariant(signature = "s")]
+// Possible values for PerformPrompt type parameter
```
Should this implement the serialisation traits or the derived ones are correct?

warusadura replied:
Yes, the derived ones are correct. I had to implement the serialisation and de-serialisation for `Reply` because of the Empty variant.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,287 @@
+    pub async fn prompt_ready(
+        &self,
+        reply: Reply,
+        _properties: Properties,
```
Why is this unused? Already mentioned it before but you marked the thread as resolved except not really...

warusadura replied:
I'll take another look.

bilelmoussaoui commented on server/src/prompt.rs:
```diff
@@ -1,25 +1,95 @@
+        #[zbus(connection)] connection: &zbus::Connection,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(), ServiceError> {
+        let callback = PrompterCallback::new(self.service.clone()).await;
```
The prompt callback creates the prompt path again, just pass it here instead

warusadura replied:
But, they are two different paths. Prompt: `/org/freedesktop/secrets/prompt/p{i}` and PrompterCallback: `/org/gnome/keyring/Prompt/p{i}`.

bilelmoussaoui commented:
cc @ueno, could you review & ack the crypto code please?

bilelmoussaoui commented on client/src/key.rs:
```diff
@@ -47,6 +47,14 @@ impl Key {
         Ok(Self::new(crypto::generate_public_key(private_key)?))
     }
 
+    pub fn generate_public_key_for_secret_exchange(
```
Those new methods needs tests, similar to the previous ones to ensure we have the same behaviour in both native & openssl crypto.

warusadura replied:
Done.

bilelmoussaoui commented on client/src/crypto/native.rs:
```diff
@@ -86,6 +86,16 @@ pub(crate) fn generate_public_key(private_key: impl AsRef<[u8]>) -> Result<Vec<u
     Ok(public_key_uint.to_bytes_be())
 }
 
+pub(crate) fn generate_public_key_for_secret_exchange(
```
As those APIs are only useful for the secret exchange, which is only used by the GNOME implementation I don't think we should have them here and instead move them to the server side under /gnome. Especially that we have a `Key::new`

warusadura replied:
Yeah, it make sense to move the secret exchange related new crypto impls to /gnome. But, we don't move the new functions in keys.rs right? Also, is it ok to have the server side as a dependency for the client side?

bilelmoussaoui replied:
I didn't say anything like that, the client is a library which cannot depend on the server which is a binary...

warusadura replied:
Sorry, I miss understood. Moved secret exchange cryto impl to gnome/crypto/.

warusadura commented on server/src/gnome/prompter.rs:
```diff
@@ -55,6 +55,24 @@ impl Properties {
+            title: Some("Unlock Keyring".to_string()),
+            choice_label: None,
+            description: Some(
+                "An application wants access to the keyring 'login', but it is locked".to_string(),
```
Yeah, fixed this thanks!

warusadura commented on server/src/service.rs:
```diff
@@ -137,8 +139,17 @@ impl Service {
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
     ) -> Result<(Vec<OwnedObjectPath>, OwnedObjectPath), ServiceError> {
-        let (unlocked, _not_unlocked) = self.set_locked(false, &objects, false).await?;
+        let (unlocked, not_unlocked) = self.set_locked(false, &objects, false).await?;
```
I have cleaned up a lot issues you mentioned and I'm looking into this now.

warusadura replied:
I don't really understand why I need to pass a Future to the Prompt and what it is suppose to do :)

bilelmoussaoui replied:
The future is an async function that would be executed once the prompt is done. So you don't have to pass the list of objects to lock/unlock to the prompt mechanism but rather the function that you would execute. by doing so, you can keep all the locking/unlocking logic in the same place and the prompting logic would be generic enough that it could be even tested on CI if needed.

warusadura replied:
Thanks for the reply, as I have mentioned [during](https://github.com/bilelmoussaoui/oo7/pull/150#discussion_r1852419482) the lock/unlock implementation, the `set_locked()` method does two things.
1: filter out and return object paths that were locked/unlocked without a prompt. We also update the locked state of an item based on the collection's locked state.
2: the actual locking/unlocking happens with a prompt (inside prompter impl). We shouldn't update the locked state right away (this is what we have now).

I agree, it does make sense (on the surface) to pass a future to the Prompt and spawn it when needed. But, during the service method Lock/Unlock `set_locked()` should return a value (objects paths that were locked/unlocked without a prompt) first. Based on that value we decide whether to issue a prompt or not. And since we already pass `service` to the Prompt, I really don't feel we need this :) The extra parameter to `set_locked()`, `has_prompt: bool` makes sure the call to `set_locked()` comes within the prompter impl based on only that we update the locked state. The locking/unlocking logic is exactly the same (generic enough).

the following is what changed (extra parameter plus) in `set_locked()` ,

```
-                    	// TODO: remove this once the prompt implementation is complete.
-                    	collection.set_locked(locked).await?;
+                    	if has_prompt {
+                        	collection.set_locked(locked).await?;
+  
```

Is passing a list of object paths to Prompt constructor a bad design?

bilelmoussaoui replied:
I haven't read any technical blockers at all but just a personal preference you have. The PR is any ways is not ready to be merged given the various unresolved yet ignored review comments.

warusadura replied:
We decided to resolve this (passing list of objects around) later: https://github.com/bilelmoussaoui/oo7/pull/173#discussion_r1976572719

bilelmoussaoui commented on server/src/gnome/keys.rs:
```diff
@@ -0,0 +1,70 @@
+use oo7::Key;
```
this file is not really useful, move it content to gnome/crypto/mod.rs

warusadura replied:
Done.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,407 @@
+    fn for_lock(object: &str) -> Self {
+        Self {
+            title: None,
+            message: Some(String::from("Lock Keyring")),
```
Use to_owned instead of String::from

warusadura replied:
Done. Does this offers any performance improvement?

bilelmoussaoui replied:
None, it is the same thing internally

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,407 @@
+            password_strength: None,
+            choice_label: None,
+            choice_chosen: None,
+            caller_window: None,
```
The caller window is something you should pass as a parameter as well

warusadura replied:
Done.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,407 @@
+    }
+}
+
+impl TryFrom<String> for Reply {
```
What uses this ?

warusadura replied:
This is to fix the following error:

```
error[E0277]: the trait bound `Reply: From<std::string::String>` is not satisfied
   --> server/src/gnome/prompter.rs:140:1
    |
140 | #[interface(name = "org.gnome.keyring.internal.Prompter.Callback")]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<std::string::String>` is not implemented for `Reply`
```

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,407 @@
+        prompt_path: OwnedObjectPath,
+        objects: Vec<OwnedObjectPath>,
+    ) -> Result<(), ServiceError> {
+        if let Some(reply) = reply.as_ref() {
```
why this and the match at the same time?

warusadura replied:
Reply (has just two variants now) is wrapped around `zvariant::Optional`. So, here if there is a value we'll match it, otherwise `else` block will handle the NoneValue which is `''`.

bilelmoussaoui replied:
Optional implements deref to Option so you should use that.

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,407 @@
+
+    pub async fn prompt(
+        &self,
+        locked: bool,
```
Why the locked here when you know the role already?

warusadura replied:
Yeah, this is set based on the role. `prompt` handles both lock & unlock.

bilelmoussaoui replied:
What i said is that you have access to the role here, the new param is useless

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,407 @@
+            password_strength: None,
+            choice_label: None,
+            choice_chosen: None,
+            caller_window: Some("".to_owned()),
```
That is not what i meant... I meant the constructor should take a param for this as well.

warusadura replied:
Done.

warusadura commented:
Hey, currently the server side openssl based tests fails with "the trait `From<ErrorStack>` is not implemented for `oo7::crypto::Error`". I think this is caused by the server side [Cargo.toml](https://github.com/bilelmoussaoui/oo7/blob/6f06a64ee8fd466e918b113ffdad6f9beafd5f53/server/Cargo.toml#L19) and I'm not sure how to fix this.

I ran the tests with: `cargo test --no-default-features --features openssl_crypto`

bilelmoussaoui commented:
> Hey, currently the server side openssl based tests fails with "the trait `From<ErrorStack>` is not implemented for `oo7::crypto::Error`". I think this is caused by the server side [Cargo.toml](https://github.com/bilelmoussaoui/oo7/blob/6f06a64ee8fd466e918b113ffdad6f9beafd5f53/server/Cargo.toml#L19) and I'm not sure how to fix this.
> 
> I ran the tests with: `cargo test --no-default-features --features openssl_crypto`

you have to enable the corresponding oo7 feature as well, in server/Cargo.toml.

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,416 @@
+                ))))
+            })?;
+            let server_exchange = secret_exchange.begin();
+            self.service
+                .set_secret_exchange_pub_key(server_exchange.clone())
+                .await;
+
```
why this information is stored in the service and not in the prompter itself? i don't think the gnome specific bits needs to be exposed in Service.

warusadura replied:
Yeah, I agree. Removed these from the Service.

bilelmoussaoui commented on server/Cargo.toml:
```diff
@@ -9,13 +9,28 @@ rust-version.workspace = true
 zeroize.workspace = true
+
+[features]
+default = ["native_crypto"]
+native_crypto = [
+    "dep:hkdf",
+    "dep:sha2",
+]
+
+openssl_crypto = ["dep:openssl"]
```
```suggestion
default = ["native_crypto"]
native_crypto = [
    "dep:hkdf",
    "dep:sha2",
    "oo7/native_crypto",
]

openssl_crypto = ["dep:openssl", "oo7/openssl_crypto"]
```

warusadura replied:
Thanks! I tried this, but I'm getting `error: You can't enable both openssl_crypto & native_crypto features at once`. Also, tried disabling oo7 default features. It didn't work either.

bilelmoussaoui replied:
How are you building it for it to fail this way?

warusadura replied:
`cargo build --no-default-features --features openssl_crypto`

warusadura replied:
Fixed this with https://github.com/bilelmoussaoui/oo7/pull/206

warusadura commented:
Hey Daiki can you please review the crypto impls (`gnome/secret_exchange.rs` and `gnome/crypto/`) when you have some free time. I just fixed an issue found in my openssl based implementation. With that everything works :) In case you wanted to run the openssl based tests you might need this [diff](https://gist.github.com/warusadura/7cc39750421f38d9bf0581758b516dc9).

@ueno

bilelmoussaoui commented on server/src/prompt.rs:
```diff
@@ -1,25 +1,97 @@
+        object_server.at(&path, callback).await?;
+        tracing::debug!("Prompt `{}` created.", self.path);
+
+        tokio::spawn(async move {
+            let prompter = PrompterProxy::new(&connection).await.unwrap();
+            prompter.begin_prompting(&path).await.unwrap();
+        });
```
Needs a comment why this has to be a separate spawned future

warusadura replied:
done.

bilelmoussaoui commented on server/src/prompt.rs:
```diff
@@ -1,25 +1,97 @@
+        &self.path
+    }
+
+    pub fn role(&self) -> &PromptRole {
+        &self.role
+    }
```
```suggestion
    pub fn role(&self) -> PromptRole {
        self.role
    }
```
it is a copy type

warusadura replied:
It's not right?

warusadura replied:
I'm getting the following error,
```
move occurs because `self.role` has type `prompt::PromptRole`, which does not implement the `Copy` trait
```
and the compiler suggest to do a clone. Should I do this or, keep it as it is?

bilelmoussaoui replied:
That is what i mentioned above.

warusadura replied:
Done, thanks!

bilelmoussaoui commented on server/src/prompt.rs:
```diff
@@ -1,25 +1,97 @@
+        &self.role
+    }
+
+    pub fn objects(&self) -> &Vec<OwnedObjectPath> {
+        &self.objects
+    }
```
```suggestion
    pub fn objects(&self) -> &[OwnedObjectPath] {
        &self.objects
    }
```
I still truly believe the objects don't make sense as part of the prompting but let us deal with that separately.

warusadura replied:
Done. Yeah I agree, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,450 @@
+        prompt_path: OwnedObjectPath,
+        objects: Vec<OwnedObjectPath>,
+    ) -> Result<(), ServiceError> {
+        let locked = match prompt_role {
+            PromptRole::Lock => true,
+            PromptRole::Unlock => false,
+            PromptRole::CreateCollection => todo!(),
+        };
```
I am not sure this is needed. The role is enough.

warusadura replied:
Sorry, I don't get it.

bilelmoussaoui replied:
I don't get what you wrote here neither. What I said is there is 0 reason to convert an enum that already conveys the information properly to a boolean

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,450 @@
+        &self.path
+    }
+
+    pub async fn object_label(&self, objects: &Vec<OwnedObjectPath>) -> Option<String> {
+        for object in objects {
+            if let Some(collection) = self.service.collection_from_path(object).await {
+                return Some(collection.label().await);
+            }
+        }
+        None
```
The objects could be items, not necessarily collections, no?

warusadura replied:
Yes. This is fixed now, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,450 @@
+        None
+    }
+
+    pub async fn set_secret_exchange_pub_key(&self, pub_key: String) {
+        *self.secret_exchange_pub_key.write().await = pub_key;
+    }
+
+    pub async fn secret_exchange_pub_key(&self) -> String {
+        self.secret_exchange_pub_key.read().await.clone()
+    }
+
+    pub async fn set_secret_exchange_aes_key(&self, aes_key: String) {
+        *self.secret_exchange_aes_key.write().await = aes_key;
+    }
+
+    pub async fn secret_exchange_aes_key(&self) -> String {
+        self.secret_exchange_aes_key.read().await.clone()
```
no need for all these methods if they are going to be called once internally. Drop them.

warusadura replied:
removed, thanks!

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,450 @@
+        self.secret_exchange_aes_key.read().await.clone()
+    }
+
+    pub async fn perform_prompt(
```
All of these static methods have 0 usefulness at all. Drop all of them.

warusadura replied:
Actually they do :) Consider `PrompterCallback::perform_prompt()` wrapper method. It's invoked in two places. So, this avoids code duplication in two places. And also with this approach I can avoid calling `.unwrap()` inside `tokio::spawn(async move` and avoid the following error, "^ cannot use the `?` operator in an async block that returns `()`"

bilelmoussaoui replied:
Well, you should make the spawned t ask return a Result instead of () https://docs.rs/tokio/latest/tokio/task/fn.spawn.html

warusadura replied:
Thanks. I think we should keep these wrapper methods, they help a lot with code duplication.

bilelmoussaoui replied:
I haven't seen any code duplication so far.

warusadura replied:
:) In `prompter.rs` file search for `stop_prompting`, `perform_prompt`, `prompt_completed` and `prompt`. Each of these static wrapper methods get executed twice.

warusadura replied:
I think this is resolved too :)

bilelmoussaoui commented:
Can you split the first two & the last commit into a separate PR, I will ping Daiki by email to review only that and we can merge it. I will provide a better approach for the secret exchange API as well so we can land that and figure out the remaining of the prompting here given that this PR is already too huge in terms of review comments.

warusadura commented:
> Can you split the first two & the last commit into a separate PR, I will ping Daiki by email to review only that and we can merge it. I will provide a better approach for the secret exchange API as well so we can land that and figure out the remaining of the prompting here given that this PR is already too huge in terms of review comments.

Yeah, done. Sorry for the delay, I've been dealing with [this](https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg2024463.html) issue in my destro.
 
I'll remove the split commits and rebase this once #213 is merged.

bilelmoussaoui commented:
This needs a rebase and please split the lock and unlock changes from the additional common changes for prompting so they can be reviewed easily

bilelmoussaoui commented on server/src/gnome/prompter.rs:
```diff
@@ -0,0 +1,445 @@
+    }
+
+    // Wrappers around PrompterProxy methods
+    pub async fn begin_prompting(
```
All of those methods are still here.

warusadura replied:
Removed all of them.

bilelmoussaoui commented:
merged most of it in https://github.com/bilelmoussaoui/oo7/pull/224, if you notice anything left/broken, patches welcome. 

Closing this one

warusadura commented:
Thanks!
